### ISSUE TITLE
pyinfra error: No such group: __file__

### ISSUE THREAD

**Issue by iredmail** [Fizzadar/pyinfra#308]
**Describe the bug**

Simple ad-hoc command failed with error.
I didn't create inventory file, just run this ad-hoc command from bash shell in any directory.

```
$ pyinfra --debug --user=root 192.168.0.115 exec -- echo
--> Loading config...
--> Loading inventory...
    [pyinfra_cli.inventory] Creating fake inventory...

--> Connecting to hosts...
    [pyinfra.api.connectors.ssh] Connecting to: 192.168.0.115 ({'allow_agent': True, 'look_for_keys': True, 'username': 'root', 'timeout': 10})
    [192.168.0.115] Connected
    [pyinfra.api.state] Activating host: 192.168.0.115
    Use of Inventory.<group_name> is deprecated, please use `Inventory.get_group` instead.
--> pyinfra error: No such group: __file__
```

**Meta**
+ Include output of `pyinfra --support`.

OS: macOS 10.15.5

```
$ pyinfra --support
--> Support information:

    If you are having issues with pyinfra or wish to make feature requests, please
    check out the GitHub issues at https://github.com/Fizzadar/pyinfra/issues .
    When adding an issue, be sure to include the following:

    System: Darwin
      Platform: Darwin-19.5.0-x86_64-i386-64bit
      Release: 19.5.0
      Machine: x86_64
    pyinfra: v0.14.5
    Executable: /usr/local/bin/pyinfra
    Python: 3.7.3 (CPython, Clang 11.0.3 (clang-1103.0.32.62))
```

+ How was pyinfra installed (source/pip)?

pip.

**Comment by pathcl** [comment]
Interesting. I don´t have a way to test it in Catalina but at least on 10.10.5 / 10.13 I cannot reproduce this issue :(

**Comment by Fizzadar** [comment]
I can't replicate this either (Catalina), unsure what's going on here.

@iredmail could you try usiing `@local`?: `pyinfra --debug @local exec -- echo`.

**Comment by iredmail** [comment]
Hi @Fizzadar

Got same error with `@local`:

```
$ pyinfra --debug @local exec -- echo.
--> Loading config...
--> Loading inventory...
    [pyinfra_cli.inventory] Creating fake inventory...

--> Connecting to hosts...
    [@local] Connected
    [pyinfra.api.state] Activating host: @local
    Use of Inventory.<group_name> is deprecated, please use `Inventory.get_group` instead.
--> pyinfra error: No such group: __file__
```

**Comment by Fizzadar** [comment]
@iredmail could you try `pyinfra --debug @local server.shell 'echo hi'`? The result is identical but it skips the `exec` path, which may be the issue here.

If that also fails I'm completely at a loss - the `__file__` seems particularly mysterious and I cannot explain it. Next step would be to try installing pyinfra in a blank virtualenv and testing that...

**Comment by iredmail** [comment]
Here it is:

```
$ pyinfra --debug @local server.shell 'echo hi'
--> Loading config...
--> Loading inventory...
    [pyinfra_cli.inventory] Creating fake inventory...

--> Connecting to hosts...
    [@local] Connected
    [pyinfra.api.state] Activating host: @local

--> Preparing operation...
    Use of Inventory.<group_name> is deprecated, please use `Inventory.get_group` instead.
--> pyinfra error: No such group: __file__
```

**Comment by Fizzadar** [comment]
Argh! So all I can think of now is to remove those methods and catch the exception - could you try the above again but after installing pyinfra from the [`temp-iredmail-test`](https://github.com/Fizzadar/pyinfra/tree/temp-iredmail-test) branch?

**Comment by iredmail** [comment]
Seems working:

```
$ pyinfra --debug @local server.shell 'echo hi'
--> Loading config...
--> Loading inventory...
    [pyinfra_cli.inventory] Creating fake inventory...

--> Connecting to hosts...
    [@local] Connected
    [pyinfra.api.state] Activating host: @local

--> Preparing operation...
    [pyinfra.api.operation] Adding operation, {'Server/Shell'}, called @ /Library/Python/3.7/site-packages/pyinfra_cli/main.py:555, opLines=(0, 555), opHash=6eab68374685829cd5f1be9723dffa823c5922a3

--> Proposed changes:
    Groups: @local
    [@local]   Operations: 1   Commands: 1   

--> Beginning operation run...
--> Starting operation: Server/Shell ('echo hi',)
    [pyinfra.api.operations] Starting operation Server/Shell on @local
    [pyinfra.api.connectors.util] Building command (shell_executable: sh): echo hi
[pyinfra.api.connectors.local] --> Running command on localhost: sh -c 'echo hi'
[pyinfra.api.connectors.util] --> Waiting for exit status...
[pyinfra.api.connectors.util] --> Command exit status: 0
    [@local] Success

--> Results:
    Groups: @local
    [@local]   Successful: 1   Errors: 0   Commands: 1/1
```

**Comment by iredmail** [comment]
It works with a remote host too.

**Comment by Fizzadar** [comment]
Interesting! I think it might be something in your Python environment - if you install `pyinfra==0.14.5` in a blank virtualenv does it work?

There's nothing changed between 0.14.5 -> master that would cause this!

**Comment by iredmail** [comment]
Seems it doesn't work:

```
(venv) ~/tmp/t/venv $ pyinfra --version
pyinfra: v0.14.5

(venv) ~/tmp/t/venv $ pyinfra --debug --user=root 192.168.0.120 exec -- echo
--> Loading config...
--> Loading inventory...
    [pyinfra_cli.inventory] Creating fake inventory...

--> Connecting to hosts...
    [pyinfra.api.connectors.ssh] Connecting to: 192.168.0.120 ({'allow_agent': True, 'look_for_keys': True, 'username': 'root', 'timeout': 10})
    [192.168.0.120] Connected
    [pyinfra.api.state] Activating host: 192.168.0.120
    Use of Inventory.<group_name> is deprecated, please use `Inventory.get_group` instead.
--> pyinfra error: No such group: __file__

(venv) ~/tmp/t/venv $ pyinfra --debug @local server.shell 'echo hi'
--> Loading config...
--> Loading inventory...
    [pyinfra_cli.inventory] Creating fake inventory...

--> Connecting to hosts...
    [@local] Connected
    [pyinfra.api.state] Activating host: @local

--> Preparing operation...
    Use of Inventory.<group_name> is deprecated, please use `Inventory.get_group` instead.
--> pyinfra error: No such group: __file__
```

**Comment by Fizzadar** [comment]
Argh! I literally can't see _anything_ in https://github.com/Fizzadar/pyinfra/compare/v0.14.5...temp-iredmail-test.

Does it work from `master` branch? My only suggestion as this point is to use master branch or upgrade to `v0.15` which I plan on releasing today.

**Comment by iredmail** [comment]
I will wait for the new release. :)

**Comment by iredmail** [comment]
Update: `master` branch failed with same error, i guess upcoming 0.15 will fail too.

**Comment by iredmail** [comment]
The error message is `No such group: __file__`, my question is: is it possible to print/save the generated inventory file? so that we can check whether it has invalid data.

**Comment by Fizzadar** [comment]
So this specifically occurs here: https://github.com/Fizzadar/pyinfra/blob/master/pyinfra/api/inventory.py#L205

This magic method was deprecated ages ago. Something is trying to do `Inventory.__file__`, which smells like some kind of Python-internal thing. BUT so weird it only appears to happen in your  environment. What's even more strange is simply removing it appears to fix it for you!

One idea - print a traceback at the top of that function? It should at least give us an idea of where this is coming from...

**Comment by iredmail** [comment]
I commented out `def __getattr_` block, then it works.
If it was deprecated ages ago, maybe it's safe to remove this block in upcoming release and make it working?

**Comment by Fizzadar** [comment]
v1 (https://github.com/Fizzadar/pyinfra/pull/232) will remove this! `v0.15` will be the last `0.x` release and I'm hoping to release v1 in the next few weeks.

I'm still keen to know how this is happening - would you mind re-running but with `import traceback; traceback.print_tb()` at the top of the function?

**Comment by iredmail** [comment]
`traceback.print_tb()` requires at least one argument. Please give me full py code snippet for this debug purpose.

**Comment by iredmail** [comment]
I added `traceback.print_stack()` right above `raise NoGroupError(...)` line, is this enough?

```
Traceback (most recent call last):
  File "/Library/Python/3.7/site-packages/pyinfra_cli/main.py", line 222, in main
    _main(*args, **kwargs)
  File "/Library/Python/3.7/site-packages/pyinfra_cli/main.py", line 555, in _main
    *args[0], **args[1]
  File "/Library/Python/3.7/site-packages/pyinfra/api/operation.py", line 71, in add_op
    frameinfo = get_caller_frameinfo()
  File "/Library/Python/3.7/site-packages/pyinfra/api/util.py", line 75, in get_caller_frameinfo
    stack_items = stack()
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/inspect.py", line 1513, in stack
    return getouterframes(sys._getframe(1), context)
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/inspect.py", line 1490, in getouterframes
    frameinfo = (frame,) + getframeinfo(frame, context)
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/inspect.py", line 1464, in getframeinfo
    lines, lnum = findsource(frame)
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/inspect.py", line 784, in findsource
    lines = linecache.getlines(file)
  File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/linecache.py", line 49, in getlines
    if getattr(mod, '__file__', None) == filename:
  File "/Library/Python/3.7/site-packages/pyinfra/pseudo_modules.py", line 21, in __getattr__
    return getattr(self._module, key)
  File "/Library/Python/3.7/site-packages/pyinfra/api/inventory.py", line 227, in __getattr__
    traceback.print_stack()
```

Seems caused by `pyinfra/pseudo_modules.py`, line 21, `return getattr(self._module, key)`?
https://github.com/Fizzadar/pyinfra/blob/master/pyinfra/pseudo_modules.py#L21

```
class PseudoModule(object):
    _module = None

    def __getattr__(self, key):
        return getattr(self._module, key)
```

**Comment by Fizzadar** [comment]
Bingo! So this appears to be specific to Apple's version of Python! - the same file is different on my Python compiled with `pyenv`! I can't explain why I've never seen this before, but I have pushed https://github.com/Fizzadar/pyinfra/commit/92b5109d0dd8becb275bd1d1351a53f107a3caae which should fix it (will be included in v0.15!).

**Comment by Fizzadar** [comment]
Just released (bit late) as `0.15.dev0` - full `0.15` will drop in a few days!

**Comment by iredmail** [comment]
0.15.dev0 works for me. Thanks @Fizzadar :)
